pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -45,87 +45,49 @@ func init() {
-		Topology:   &Topology{kubeClient: kubeClient},
+		kubeClient: kubeClient,
+		topology:   &Topology{kubeClient: kubeClient},
 	}
 }
 
-func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) ([]*Schedule, error) {
+func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) ([]*TheoreticalNode, error) {
 	defer metrics.Measure(schedulingDuration.WithLabelValues(injection.GetNamespacedName(ctx).Name))()
 	constraints := provisioner.Spec.Constraints.DeepCopy()
 	// Inject temporarily adds specific NodeSelectors to pods, which are then
 	// used by scheduling logic. This isn't strictly necessary, but is a useful
 	// trick to avoid passing topology decisions through the scheduling code. It
-	// lets us to treat TopologySpreadConstraints as just-in-time NodeSelectors.
-	if err := s.Topology.Inject(ctx, constraints, pods); err != nil {
+	// lets us treat TopologySpreadConstraints as just-in-time NodeSelectors.
+	if err := s.topology.Inject(ctx, constraints, pods); err != nil {
 		return nil, fmt.Errorf("injecting topology, %w", err)
 	}
-	// Separate pods into schedules of isomorphic scheduling constraints.
-	return s.getSchedules(constraints, instanceTypes, pods), nil
+	return s.schedule(constraints, instanceTypes, pods), nil
 }
 
-// getSchedules separates pods into a set of schedules. All pods in each group
-// contain isomorphic scheduling constraints and can be deployed together on the
-// same node, or multiple similar nodes if the pods exceed one node's capacity.
-func (s *Scheduler) getSchedules(constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) []*Schedule {
-	schedules := []*Schedule{}
+// schedule separates pods into a list of TheoreticalNodes that contain the pods. All pods in each theoretical node
+// contain isomorphic scheduling constraints and can be deployed together on the same node, or multiple similar nodes if
+// the pods exceed one node's capacity.
+func (s *Scheduler) schedule(constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) []*TheoreticalNode {
+	var nodes []*TheoreticalNode
 	for _, pod := range pods {
-		isCompatible := false
-		for _, schedule := range schedules {
-			if err := schedule.Requirements.Compatible(v1alpha5.NewPodRequirements(pod)); err == nil {
-				// Test if there is any instance type that can support the combined constraints
-				// TODO: Implement a virtual node approach solution that combine scheduling and node selection to solve this problem.
-				c := schedule.Tighten(pod)
-				for _, instanceType := range instanceTypes {
-					if support(instanceType, c) {
-						schedule.Constraints = c
-						schedule.Pods = append(schedule.Pods, pod)
-						isCompatible = true
-						break
-					}
-				}
+		canScheduleOnExistingNode := false
```

Comment:
Optional more brief name
```suggestion
		isScheduled := false
```

Reply:
SGTM